File: src/main/scala/dregex/extra/DotFormatter.scala
package dregex.extra

import dregex.impl.Epsilon
import dregex.impl.Nfa

object DotFormatter {

  def format(nfa: Nfa): String = {
    val states = for (state <- nfa.allStates) yield {
      val shape =
        if (state == nfa.initial)
          "square"
        else
          "circle"
      val peripheries =
        if (nfa.accepting.contains(state))
          2
        else
          1
      s""""${state.toString}" [shape=$shape,peripheries=$peripheries];"""
    }
    val transitions = for (transition <- nfa.transitions) yield {
      val weight =
        if (transition.char == Epsilon.instance)
          1
        else
          2
      s""""${transition.from.toString}" -> "${transition.to.toString}" [label=${transition.char.toString}, weight=$weight];"""
    }
    s"""
      digraph graphname {
        rankdir=LR;
        ${states.mkString("\n")}
        ${transitions.mkString("\n")}
      }
    """
  }

}

File: src/test/java/dregex/extra/DotFormatter.java
package dregex.extra;

import dregex.impl.Epsilon;
import dregex.impl.Nfa;

import java.util.stream.Collectors;

class DotFormatter {

    public static String format(Nfa nfa) {
        var states = nfa.collectAllStates().stream().map(state -> {
            String shape;
            if (state.equals(nfa.initial)) {
                shape = "square";
            } else {
                shape = "circle";
            }
            int peripheries;
            if (nfa.accepting.contains(state)) {
                peripheries = 2;
            } else {
                peripheries = 1;
            }
            return String.format("\"%s\"[shape = %s, peripheries = %s];", state, shape, peripheries);
        });
        var transitions = nfa.transitions.stream().map(transition -> {
            int weight;
            if (transition.ch == Epsilon.instance) {
                weight = 1;
            } else {
                weight = 2;
            }
            return String.format("\"%s\" -> \"%s\" [label=%s, weight=%s];", transition.from, transition.to, transition.ch, weight);
        });
        return String.format("digraph graphname { rankdir=LR; %s %s}",
                states.collect(Collectors.joining("\n")), transitions.collect(Collectors.joining("\n")));
    }

}
